internal/kafka/internal/workers/kafka_mgrs/kafkas_mgr.go

```diff
@@ -343,3 +355,95 @@ func (k *KafkaManager) updateClusterStatusCapacityMaxMetric(c services.KafkaStre
 func (k *KafkaManager) updateZeroValueOfKafkaRequestsExpiredAt() error {
 	return k.kafkaService.UpdateZeroValueOfKafkaRequestsExpiredAt()
 }
+
+func (k *KafkaManager) reconcileKafkaExpiresAt(kafkas dbapi.KafkaList) serviceErr.ErrorList {
+	var svcErrors serviceErr.ErrorList
+	subscriptionStatusByOrgAndBillingModel := map[string]bool{}
+
+	for _, kafka := range kafkas {
+		// skip update when Kafka is marked for deletion or is already being deleted
+		if arrays.Contains(constants.GetDeletingStatuses(), kafka.Status) {
+			continue
+		}
+
+		instanceSize, err := k.kafkaConfig.GetKafkaInstanceSize(kafka.InstanceType, kafka.SizeId)
+		if err != nil {
+			svcErrors = append(svcErrors, errors.Wrapf(err,
+				"failed to get kafka instance size for %q with instance type %q and size id %q",
+				kafka.ID, kafka.InstanceType, kafka.SizeId,
+			))
+			continue
+		}
+
+		expiresAt := kafka.ExpiresAt
+		// If lifespan seconds is not defined, expiration is determined based on the organisation's active subscription
+		if instanceSize.LifespanSeconds == nil {
+			orgBillingModelID := fmt.Sprintf("%s-%s", kafka.OrganisationId, kafka.ActualKafkaBillingModel)
+			active, ok := subscriptionStatusByOrgAndBillingModel[orgBillingModelID]
+			if !ok {
+				isActive, err := k.kafkaService.IsSubscriptionActive(kafka)
+				if err != nil {
+					svcErrors = append(svcErrors, errors.Wrapf(err, "failed to get subscription status of kafka instance %q", kafka.ID))
+					continue
+				}
+				subscriptionStatusByOrgAndBillingModel[orgBillingModelID] = *isActive
+				active = *isActive
+			}
+
+			if active {
+				// if subscription is active, set expires_at to null
+				if err := expiresAt.Scan(nil); err != nil {
+					svcErrors = append(svcErrors, errors.Wrapf(err, "failed to set expires_at value for kafka instance %q to NULL", kafka.ID))
+					continue
+				}
+			} else {
+				// if subscription is not valid and expires_at is not already set, set its value based on the current time and grace period allowance
+				if !kafka.ExpiresAt.Valid {
+					// set expires_at to now + grace period days
+					billingModel, err := k.kafkaConfig.GetBillingModelByID(kafka.InstanceType, kafka.ActualKafkaBillingModel)
+					if err != nil {
+						svcErrors = append(svcErrors, errors.Wrapf(err, "failed to get billing model of kafka instance %q", kafka.ID))
+						continue
+					}
+
+					duration, err := time.ParseDuration(fmt.Sprintf("%dh", billingModel.GracePeriodDays*24))
```

Comment:
Same comment as https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1499/files#r1086451249 but using different units.

For example:
```
duration := time.Duration(gracePeriodDays*86400) * time.Second
```